src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.Unix.cs

```diff
@@ -349,14 +383,30 @@ public static int Wait(
             Debug.Assert(timeoutMilliseconds >= -1);
 
             ThreadWaitInfo waitInfo = Thread.CurrentThread.WaitInfo;
+
+#if FEATURE_CROSS_PROCESS_MUTEX
+            if (waitHandles.Length == 1 && HandleManager.FromHandle(waitHandles[0]) is NamedMutex namedMutex)
+            {
+                // Named mutexes don't participate in the wait subsystem fully.
+                return namedMutex.Wait(waitInfo, timeoutMilliseconds, interruptible: true, prioritize: false);
+            }
+#endif
+
             WaitableObject?[] waitableObjects = waitInfo.GetWaitedObjectArray(waitHandles.Length);
             bool success = false;
+
             try
             {
                 for (int i = 0; i < waitHandles.Length; ++i)
                 {
                     Debug.Assert(waitHandles[i] != IntPtr.Zero);
-                    WaitableObject waitableObject = HandleManager.FromHandle(waitHandles[i]);
+                    IWaitableObject waitableObjectMaybe = HandleManager.FromHandle(waitHandles[i]);
+
+                    if (waitableObjectMaybe is not WaitableObject waitableObject)
+                    {
+                        throw new ArgumentException("Only unnamed waitable objects are supported in multi-wait operations.", nameof(waitHandles));
```

Comment:
We have existing error message and exception for this situation:

System.PlatformNotSupportedException: Wait operations on multiple wait handles including a named synchronization primitive are not supported on this platform.

Also, we should have a test to cover this case.

Reply:
I'll work on adding tests for various corner cases.